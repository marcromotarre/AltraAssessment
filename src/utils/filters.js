export const isMinMaxFilterApplied = (filter) =>
  filter.min === '' && filter.max === '' ? false : true;

export const isArrayFilterApplied = (filter) =>
  filter.length > 0 ? true : false;

export const getFilterMinMaxString = (data) => {
  if (data.range && data.min !== '' && data.max !== '') {
    return `${data.min} - ${data.max}`;
  } else if (data.min !== '') {
    return `${data.min}`;
  } else if (data.max !== '') {
    return `${data.max}`;
  } else {
    return '';
  }
};

export const isAgeOrWeightOrHeightFilterApplied = ({ age, height, weight }) => {
  return isMinMaxFilterApplied(age) ||
    isMinMaxFilterApplied(height) ||
    isMinMaxFilterApplied(weight)
    ? true
    : false;
};

export const isAnyFilterApplied = ({ age, height, weight, hair_colors }) => {
  return isAgeOrWeightOrHeightFilterApplied({ age, height, weight }) ||
    isArrayFilterApplied(hair_colors)
    ? true
    : false;
};

export const getUsersByName = (users, name) => {
  return users.filter((user) =>
    user.name.toLowerCase().includes(name.toLowerCase())
  );
};

export const getUsersByProfession = (users, profession) => {
  return users.filter((user) => user.professions.includes(profession));
};

export const getUniqueHairColors = (users) => {
  return users
    .map((user) => user.hair_color)
    .filter((value, index, self) => self.indexOf(value) === index);
};

export const getUsersByMinMaxFilter = (users, type, filter) => {
  if (!isMinMaxFilterApplied(filter)) {
    return [...users];
  } else {
    if (filter.min !== '' && filter.max !== '' && filter.range) {
      return users.filter((user) => {
        return (
          user[type] >= parseFloat(filter.min) &&
          user[type] <= parseFloat(filter.max)
        );
      });
    } else if (
      (filter.min !== '' && filter.max !== '' && !filter.range) ||
      filter.min !== ''
    ) {
      return users.filter((user) => {
        return Math.round(user[type]) === parseFloat(filter.min);
      });
    } else {
      return users.filter((user) => {
        return Math.round(user[type]) === parseFloat(filter.max);
      });
    }
  }
};

export const getUsersByArrayFilter = (users, type, filter) => {
  if (!isArrayFilterApplied(filter)) {
    return [...users];
  } else {
    return users.filter((user) => {
      return filter.includes(user[type]);
    });
  }
};

export const getUsersByFilters = ({
  users = [],
  age = { min: '', max: '', range: false },
  height = { min: '', max: '', range: false },
  weight = { min: '', max: '', range: false },
  hair_colors = [],
}) => {
  return getUsersByArrayFilter(
    getUsersByMinMaxFilter(
      getUsersByMinMaxFilter(
        getUsersByMinMaxFilter(users, 'age', age),
        'weight',
        weight
      ),
      'height',
      height
    ),
    'hair_color',
    hair_colors
  );
};
