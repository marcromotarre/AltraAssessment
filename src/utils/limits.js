export const getMinUserValueByField = (users, field) => {
  return users ? Math.min(...users.map((user) => Math.floor(user[field]))) : 0;
};

export const getMaxUserValueByField = (users, field) => {
  return users ? Math.max(...users.map((user) => Math.ceil(user[field]))) : 0;
};
