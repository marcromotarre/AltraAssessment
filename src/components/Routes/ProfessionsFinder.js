import React, { useState } from 'react';
import styled from 'styled-components';
import { useSelector } from 'react-redux';

import { useParams } from '@reach/router';

import ProfessionsSVG from './../../svg/professions.svg';
import { getUsersByName, getUsersByProfession } from '../../utils/filters.js';

import Profession from '../Profession';
import SectionBanner from '../Banners/SectionBanner';
import UsersList from '../User/UsersList.js';
import SearchInput from '../SearchInput.js';
import DataManager from '../DataManager';

export const ProfessionHeader = styled.div`
  background: white;
  position: sticky;
  top: 0;
  display: grid;
  grid-template-columns: 100%;
  row-gap: 13px;

  svg,
  section {
    justify-self: center;
  }
`;

const ProfessionsFinder = () => {
  let { name } = useParams();
  const { users } = useSelector((state) => state.town);
  const [searchValue, setSearchValue] = useState('');

  const usersByProfession = getUsersByProfession(users, name);
  const searchedUsers = getUsersByName(usersByProfession, searchValue);

  const search = (inputSearchValue) => {
    setSearchValue(inputSearchValue);
  };

  return (
    <section data-testid="ProfessionSite">
      <ProfessionHeader>
        <SectionBanner />
        <ProfessionsSVG width={49.83}> </ProfessionsSVG>
        <Profession name={name}> </Profession>
        <SearchInput onSearch={search} />
      </ProfessionHeader>
      <DataManager>
        <UsersList users={searchedUsers} />
      </DataManager>
    </section>
  );
};

export default ProfessionsFinder;
