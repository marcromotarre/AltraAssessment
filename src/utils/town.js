import { getUniqueHairColors } from './filters';

export const getTownsNames = (data) => Object.keys(data);

export const getUsersByTown = (data, townName) => data[townName];

export const getUniqueHairColorsByTown = (data, townName) =>
  getUniqueHairColors(data[townName]);
