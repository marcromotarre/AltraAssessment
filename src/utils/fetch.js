import fetch from 'cross-fetch';
import * as actions from './../redux/actions/town';

export function fetchTown(dispatch) {
  dispatch(actions.fetchTownPending());
  fetch(
    'https://raw.githubusercontent.com/rrafols/mobile_test/master/data.json'
  )
    .then((response) => response.json())
    .then((response) => {
      if (response.error) {
        throw response.error;
      }
      dispatch(actions.fetchTownSuccess(response));
    })
    .catch((error) => {
      dispatch(actions.fetchTownError(error));
    });
}
