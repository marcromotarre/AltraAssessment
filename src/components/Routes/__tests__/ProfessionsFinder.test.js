import React from 'react';
import { renderRouteRedux } from '../../../utils/test';
import users from '../../../fixtures/users';
import '@testing-library/jest-dom/extend-expect';

import { Router } from '@reach/router';

import ProfessionsFinder from '../ProfessionsFinder';

describe('User Route Component', () => {
  it('User Route Snapshot', () => {
    expect(
      renderRouteRedux(
        <Router>
          <ProfessionsFinder path="/profession/:name" />
        </Router>,
        {
          route: '/profession/profession_1',
          initialState: { town: { users: users } },
        }
      ).baseElement
    ).toMatchSnapshot();
  });
});
