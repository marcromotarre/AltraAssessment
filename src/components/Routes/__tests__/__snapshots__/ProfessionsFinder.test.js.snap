// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Route Component User Route Snapshot 1`] = `
<body>
  <div>
    <div
      style="outline: none;"
      tabindex="-1"
    >
      <section
        data-testid="ProfessionSite"
      >
        <div
          class="sc-fzokOt khFfoJ"
        >
          <div
            class="sc-AxirZ eEbsrJ"
          >
            <a
              href="/"
            >
              <svg
                width="25"
              />
            </a>
          </div>
          <svg
            width="49.83"
          >
             
          </svg>
          <section
            class="sc-AxjAm hHoNSr"
          >
            <label>
              profession_1
            </label>
          </section>
          <div
            class="sc-fzpans iRGLvn true"
          >
            <svg
              class="sc-fzplWN hZJcUG"
              data-testid="search-icon"
              width="21.42"
            />
            <input
              class="sc-fznyAO hVkSWH"
              data-testid="search-input"
              value=""
            />
          </div>
        </div>
        <div
          class="sc-fzoLsD bLrCsw"
        >
          <a
            href="/user/0"
          >
            <section
              class="sc-AxhUy ikKUnn"
            >
              <img
                class="sc-AxgMl kIBvbX"
                data-testid="user-image"
                src="image1"
              />
              <div
                class="sc-AxheI dJYnZY"
              >
                <label
                  class="sc-Axmtr gjpNkM"
                  data-testid="user-name"
                >
                  one
                </label>
                <div
                  class="sc-AxmLO iTqPkU"
                  data-testid="user-properties"
                >
                  <div
                    class="sc-AxiKw bdxaEw"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.63"
                    />
                     
                    <label>
                      10
                    </label>
                  </div>
                  <div
                    class="sc-AxiKw bdxaEw"
                    data-testid="user-property"
                  >
                    <svg
                      height="17.7"
                    />
                    <label>
                      50
                    </label>
                  </div>
                  <div
                    class="sc-AxiKw bdxaEw"
                    data-testid="user-property"
                  >
                    <svg
                      width="17.7"
                    />
                    <label>
                      40
                    </label>
                  </div>
                  <div
                    class="sc-AxiKw bdxaEw"
                    data-testid="user-property"
                  >
                    <svg
                      class="sc-AxhCb ibBHog"
                      color="Red"
                      width="15.26"
                    />
                  </div>
                  <div
                    class="sc-AxiKw bdxaEw"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.34"
                    />
                    <label>
                      3
                    </label>
                  </div>
                  <div
                    class="sc-AxiKw bdxaEw"
                    data-testid="user-property"
                  >
                    <svg
                      width="15.26"
                    />
                     
                    <label>
                      2
                    </label>
                  </div>
                </div>
              </div>
            </section>
          </a>
          <a
            href="/user/2"
          >
            <section
              class="sc-AxhUy ikKUnn"
            >
              <img
                class="sc-AxgMl kIBvbX"
                data-testid="user-image"
                src="image3"
              />
              <div
                class="sc-AxheI dJYnZY"
              >
                <label
                  class="sc-Axmtr gjpNkM"
                  data-testid="user-name"
                >
                  three
                </label>
                <div
                  class="sc-AxmLO iTqPkU"
                  data-testid="user-properties"
                >
                  <div
                    class="sc-AxiKw bdxaEw"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.63"
                    />
                     
                    <label>
                      40
                    </label>
                  </div>
                  <div
                    class="sc-AxiKw bdxaEw"
                    data-testid="user-property"
                  >
                    <svg
                      height="17.7"
                    />
                    <label>
                      120
                    </label>
                  </div>
                  <div
                    class="sc-AxiKw bdxaEw"
                    data-testid="user-property"
                  >
                    <svg
                      width="17.7"
                    />
                    <label>
                      100
                    </label>
                  </div>
                  <div
                    class="sc-AxiKw bdxaEw"
                    data-testid="user-property"
                  >
                    <svg
                      class="sc-AxhCb ibBHog"
                      color="Red"
                      width="15.26"
                    />
                  </div>
                  <div
                    class="sc-AxiKw bdxaEw"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.34"
                    />
                    <label>
                      2
                    </label>
                  </div>
                  <div
                    class="sc-AxiKw bdxaEw"
                    data-testid="user-property"
                  >
                    <svg
                      width="15.26"
                    />
                     
                    <label>
                      2
                    </label>
                  </div>
                </div>
              </div>
            </section>
          </a>
        </div>
      </section>
    </div>
  </div>
</body>
`;
